File: core/src/main/java/org/elkd/core/consensus/RaftFollowerDelegate.java
package org.elkd.core.consensus;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import io.grpc.stub.StreamObserver;
import org.apache.log4j.Logger;
import org.elkd.core.config.Config;
import org.elkd.core.consensus.messages.AppendEntriesRequest;
import org.elkd.core.consensus.messages.AppendEntriesResponse;
import org.elkd.core.consensus.messages.Entry;
import org.elkd.core.consensus.messages.RequestVoteRequest;
import org.elkd.core.consensus.messages.RequestVoteResponse;
import org.elkd.core.log.LogChangeReason;
import org.elkd.core.log.commands.AppendFromCommand;
import org.elkd.core.log.commands.CommitCommand;

import javax.annotation.Nonnull;

class RaftFollowerDelegate implements RaftState {
  private static final Logger LOG = Logger.getLogger(RaftFollowerDelegate.class.getName());

  private final Raft mRaft;

  private TimeoutMonitor mTimeoutMonitor;

  RaftFollowerDelegate(@Nonnull final Raft raft) {
    this(
        raft,
        new TimeoutMonitor(
            raft.getConfig().getAsInteger(Config.KEY_RAFT_ELECTION_TIMEOUT_MS),
            () -> raft.transitionToState(RaftCandidateDelegate.class)
        )
    );
  }

  @VisibleForTesting
  RaftFollowerDelegate(@Nonnull final Raft raft,
                       @Nonnull final TimeoutMonitor timeoutMonitor) {
    mRaft = Preconditions.checkNotNull(raft, "raft");
    mTimeoutMonitor = Preconditions.checkNotNull(timeoutMonitor, "timeoutMonitor");
  }

  @Override
  public void on() {
    LOG.info("ready");
    mTimeoutMonitor.reset();
  }

  @Override
  public void off() {
    LOG.info("offline");
    mTimeoutMonitor.stop();
  }

  @Override
  public void delegateAppendEntries(final AppendEntriesRequest appendEntriesRequest,
                                    final StreamObserver<AppendEntriesResponse> responseObserver) {
    try {
      validateAppendEntriesRequest(appendEntriesRequest);
      final AppendFromCommand<Entry> appendFromCommand = AppendFromCommand.Companion
          .build(appendEntriesRequest.getPrevLogIndex() + 1, appendEntriesRequest.getEntries(), LogChangeReason.REPLICATION);
      mRaft.getLogCommandExecutor().execute(appendFromCommand);

      if (appendEntriesRequest.getLeaderCommit() > mRaft.getLog().getCommitIndex()) {
        final long commitIndex = Math.min(appendEntriesRequest.getLeaderCommit(), mRaft.getLog().getLastIndex());
        mRaft.getLogCommandExecutor().execute(CommitCommand.build(commitIndex, LogChangeReason.REPLICATION));
      }
      replySuccess(responseObserver);
    } catch (final Exception e) {
      replyFalse(responseObserver);
    } finally {
      mTimeoutMonitor.reset();
    }
  }

  @Override
  public void delegateRequestVote(final RequestVoteRequest requestVoteRequest,
                                  final StreamObserver<RequestVoteResponse> responseObserver) {
    mTimeoutMonitor.reset();
    responseObserver.onCompleted();
  }

  private void replySuccess(final StreamObserver<AppendEntriesResponse> responseObserver) {
    responseObserver.onNext(AppendEntriesResponse.builder(mRaft.getRaftContext().getCurrentTerm(), true).build());
    responseObserver.onCompleted();
  }

  private void validateAppendEntriesRequest(final AppendEntriesRequest appendEntriesRequest) throws Exception {
    if (appendEntriesRequest.getTerm() < mRaft.getRaftContext().getCurrentTerm()) {
      throw new Exception("Term mismatch. Validation Failed.");
    }
    /* This check is only relevant when there are entries in the log.
       `prevLogIndex == -1` occurs for the first ever log entry.
     */
    if (appendEntriesRequest.getPrevLogIndex() > -1) {
      final Entry prevEntry = mRaft.getLog().read(appendEntriesRequest.getPrevLogIndex());
      if (prevEntry == null || (prevEntry.getTerm() != appendEntriesRequest.getPrevLogTerm())) {
        throw new Exception("Entry.term mismatch. Validation Failed.");
      }
    }
  }

  /*
   * Informs sender that the entries were not appended to the log.
   */
  private void replyFalse(final StreamObserver<AppendEntriesResponse> responseObserver) {
    responseObserver.onNext(AppendEntriesResponse.builder(mRaft.getRaftContext().getCurrentTerm(), false).build());
    responseObserver.onCompleted();
  }
}

File: core/src/main/java/org/elkd/core/consensus/RaftFollowerDelegate.kt
package org.elkd.core.consensus

import com.google.common.annotations.VisibleForTesting
import io.grpc.stub.StreamObserver
import org.apache.log4j.Logger
import org.elkd.core.config.Config
import org.elkd.core.consensus.messages.AppendEntriesRequest
import org.elkd.core.consensus.messages.AppendEntriesResponse
import org.elkd.core.consensus.messages.RequestVoteRequest
import org.elkd.core.consensus.messages.RequestVoteResponse
import org.elkd.core.log.LogChangeReason
import org.elkd.core.log.commands.AppendFromCommand
import org.elkd.core.log.commands.CommitCommand
import org.elkd.shared.annotations.Mockable
import org.elkd.shared.math.randomizeNumberPoint
import kotlin.math.min

@Mockable
class RaftFollowerDelegate @VisibleForTesting
constructor(private val raft: Raft,
            private val timeoutMonitor: TimeoutMonitor) : RaftState {
  private val timeout = raft.config.getAsInteger(Config.KEY_RAFT_FOLLOWER_TIMEOUT_MS)

  constructor(raft: Raft) : this(
      raft,
      TimeoutMonitor { raft.transition(RaftCandidateDelegate::class.java) }
  )

  override fun on() {
    LOG.info("follower ready")
    resetTimeout()
  }

  override fun off() {
    LOG.info("follower offline")
    timeoutMonitor.stop()
  }

  override fun delegateAppendEntries(appendEntriesRequest: AppendEntriesRequest,
                                     responseObserver: StreamObserver<AppendEntriesResponse>) {
    try {
      LOG.info(raft.log)
      validateAppendEntriesRequest(appendEntriesRequest)
      val appendFromCommand = AppendFromCommand
          .build(appendEntriesRequest.prevLogIndex + 1, appendEntriesRequest.entries, LogChangeReason.REPLICATION)
      raft.logCommandExecutor.execute(appendFromCommand)

      commitIfNecessary(appendEntriesRequest)
      replyAppendEntries(true, responseObserver)
    } catch (e: Exception) {
      LOG.error(e)
      replyAppendEntries(false, responseObserver)
    } finally {
      resetTimeout()
    }
  }

  private fun commitIfNecessary(appendEntriesRequest: AppendEntriesRequest) {
    if (appendEntriesRequest.leaderCommit > raft.log.commitIndex) {
      val commitIndex = min(appendEntriesRequest.leaderCommit, raft.log.lastIndex)
      raft.logCommandExecutor.execute(CommitCommand.build(commitIndex, LogChangeReason.REPLICATION))
    }
  }

  override fun delegateRequestVote(requestVoteRequest: RequestVoteRequest,
                                   responseObserver: StreamObserver<RequestVoteResponse>) {
    val raftContext = raft.raftContext
    val currentTerm = raftContext.currentTerm

    if (requestVoteRequest.term < currentTerm) {
      replyRequestVote(false, responseObserver)
      return
    }

    if (raftContext.votedFor in listOf(null, requestVoteRequest.candidateId)
        && raft.log.lastIndex <= requestVoteRequest.lastLogIndex
        && raft.log.lastEntry.term <= requestVoteRequest.lastLogTerm) {
      raftContext.votedFor = requestVoteRequest.candidateId
      raftContext.currentTerm = requestVoteRequest.term
      replyRequestVote(true, responseObserver)
      return
    }

    replyRequestVote(false, responseObserver)
    return
  }

  @Throws(Exception::class)
  private fun validateAppendEntriesRequest(appendEntriesRequest: AppendEntriesRequest) {
    if (appendEntriesRequest.term < raft.raftContext.currentTerm) {
      throw Exception("Term mismatch. Validation Failed: requestTerm: ${appendEntriesRequest.term}, currentTerm:${raft.raftContext.currentTerm}")
    }

    val prevEntry = raft.log.read(appendEntriesRequest.prevLogIndex)
    if (prevEntry != null && prevEntry.term != appendEntriesRequest.prevLogTerm) {
      throw Exception("Entry.term mismatch. Validation Failed: prevLogIndex: ${appendEntriesRequest.prevLogIndex}, request: ${appendEntriesRequest.prevLogTerm}, prevEntry: ${prevEntry?.term}")
    }
  }

  private fun replyAppendEntries(response: Boolean, responseObserver: StreamObserver<AppendEntriesResponse>) {
    responseObserver.onNext(AppendEntriesResponse.builder(raft.raftContext.currentTerm, response).build())
    responseObserver.onCompleted()
  }

  private fun replyRequestVote(response: Boolean, responseObserver: StreamObserver<RequestVoteResponse>) {
    responseObserver.onNext(RequestVoteResponse.builder(raft.raftContext.currentTerm, response).build())
    responseObserver.onCompleted()
  }

  private fun resetTimeout() {
    val newTimeout = randomizeNumberPoint(timeout, 0.2)
    LOG.info("timeout in " + newTimeout + "ms")
    timeoutMonitor.reset(newTimeout.toLong())
  }

  companion object {
    private val LOG = Logger.getLogger(RaftFollowerDelegate::class.java.name)
  }
}
